lib/b3fs/src/bucket/dir/reader.rs

```diff
@@ -1,35 +1,347 @@
+        let key_phf_table = 8;
+        let disps_len = (num_entries as f64 / 5.0).ceil() as usize * 4;
+        let entries_len = num_entries as usize * 4;
+        let length_phf_table = key_phf_table + disps_len + entries_len;
+        let position_start_entries =
+            position_start_bloom_filter + length_bloom_filter as u64 + length_phf_table as u64;
         Self {
             num_entries: num_entries as u16,
             file,
+            bloom_filter: OnceCell::new(),
+            hashtree: OnceCell::new(),
+            position_start_hashes,
+            position_start_bloom_filter,
+            position_start_entries,
+            position_start_phf_table,
+            phf_table: OnceCell::new(),
+        }
+    }
+
+    pub async fn hashtree(&mut self) -> Result<&HashTree<'static>, errors::ReadError> {
+        self.hashtree
+            .get_or_try_init(|| async {
+                let mut buffer = Vec::with_capacity(self.num_entries as usize * 32);
+                let mut file_reader = self.position_file(self.position_start_hashes).await?;
+                file_reader.read_exact(&mut buffer).await?;
+                let boxed_slice = buffer.into_boxed_slice();
+                let static_slice: &'static [u8] = Box::leak(boxed_slice);
+                HashTree::try_from(static_slice).map_err(|e| errors::ReadError::InvalidHashtree(e))
+            })
+            .await
+    }
+
+    pub async fn get_entry<'a>(
+        &'a mut self,
+        name: &'a [u8],
+    ) -> Result<Option<BorrowedEntry<'a>>, errors::ReadError> {
+        let bloom_filter = self.get_bloom_filter().await?;
+        if bloom_filter.contains(name) {
+            let phf_table = self.get_phf_table().await?;
+            let entry_offset = self.get_entry_offset(name, phf_table)?;
+            return self.read_entry_at_offset(name, entry_offset).await;
         }
+        Ok(None)
+    }
+
+    pub async fn entries<'a>(&'a self) -> Result<DirEntriesIter<'a>, errors::ReadError> {
+        let result =
+            DirEntriesIter::new(self.file.try_clone().await?, self.position_start_entries).await?;
+        Ok(result)
+    }
+
+    async fn position_file(&self, position: u64) -> Result<fs::File, errors::ReadError> {
+        let mut file = self
+            .file
+            .try_clone()
+            .await
+            .map_err(|_| errors::ReadError::RefFile)?;
+        file.seek(SeekFrom::Start(position))
+            .await
+            .map_err(|_| errors::ReadError::RefFile)?;
+        Ok(file)
+    }
+
+    async fn get_bloom_filter(&self) -> Result<&BloomFilter, errors::ReadError> {
+        let result = self
+            .bloom_filter
+            .get_or_try_init(|| async {
+                let mut file = self.position_file(self.position_start_bloom_filter).await?;
+                let hashes_bloom_filter_len = file.read_u32_le().await?;
+
+                let bloom_filter_len = self.num_entries as usize * 4;
```

Comment:
this is different that the way you compute the `bloom_filter_len` during the `new` function.

Reply:
I changed that part for a centralized HeaderPositions struct to avoid this issue between reader and writer